drivers: display: add support for RM68200 display to RT595

danieldegrasse opened the pull request:
Add support for the RM68200 display with the RT595 EVK. This PR includes the following new drivers:
- LCDIF display driver for RT595 LCDIF IP
- MIPI DSI driver for RT595 MIPI IP block

In addition, the following drivers have been updated:
- ~RM68200 driver now manages power pin directly, so that regulator API is not required to manage panel power~
- FlexSPI MEMC driver now permits reserving AHB RX buffer space for specific AHB masters. This can greatly improve performance
- support for querying LCDIF clock rates has been added to LPC syscon clock control driver
- GT911 Kscan driver has been updated to correctly support using the GT911 without interrupts (in polling mode)

This PR was tested with the LVGL widgets sample, `samples/drivers/display`, and `samples/subsys/display/lvgl` on the RT595 and RT1170 EVK using the RK055HDMIPI4M display

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: memc_mcux_flexspi: enable configuring AHB RX buffer allocation

Allow configuration of AHB RX buffer allocation. This allows sections
of the AHB RX buffer to be reserved for specific masters, which can
enhance performance.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: mipi: introduce MCUX MIPI DSI 2L driver

Introduce driver for MCUX MIPI DSI 2L. This IP block differs slightly from
the existing MCUX MIPI peripheral, and uses a different hardware
abstraction layer. For these reasons, a new driver was introduced rather
than extending the existing mcux_dsi implementation.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: display: introduce driver for NXP DCNANO LCDIF peripheral

Introduce driver for NXP DCNANO LCDIF (lcd interface) peripheral,
present on iMX.RT500. Currently this driver only supports updating
the primary framebuffer, and does not implement support for the cursor
buffer present on this IP.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: clock_control: add clock rate definitions for MIPI and LCDIF

Add clock rate definitions for MIPI and LCDIF peripherals, to enable
retrival of these peripheral clock rates at runtime.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp_rt5xx: add MIPI and LCDIF nodes

Add MIPI and LCDIF node definitions, including clock devices for
the MIPI controller.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: arm: nxp_imx: r5xx: add clock initialization for MIPI and LCDIF

Add clock initialization for MIPI and LCIDF to NXP RT5xx SOC.
Note that clock divider properties are used by both initialization
routines, as the required clock divider will vary depending on
attached display.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: arm: mimxrt595_evk: add configuration for RM68200 display

Add configuration for RM68200 MIPI display. This display
controller drives a 1280x720 LCD display.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
manifest: update NXP HAL revision

Update revision of NXP HAL to add support for MIPI DSI and MCUX LCDIF
drivers.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: drivers: display: add support for RT595 EVK

Add support for RT595 EVK to display sample using RM68200 based MIPI
panel. The board is configured with an increased heap size to account
for the panel's high resolution.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

DerekSnell commented on dts/bindings/display/nxp,lcdif.yaml:
```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2022, NXP
```
```suggestion
# Copyright (c) 2023, NXP
```

DerekSnell commented on dts/bindings/display/nxp,lcdif.yaml:
```diff
@@ -0,0 +1,126 @@
+      - "unused"
+      - "xrgb-4444" # XRGB4444, 16-bit each pixel, 4-bit each element. R4G4B4 in reference manual.
+      - "xrgb-1555" # XRGB1555, 16-bit each pixel, 5-bit each element. R5G5B5 in reference manual.
+      - "rgb-565" # RGB565, 16-bit each pixel. R5G6B5 in reference manual.
```
nit, align comment with others

danieldegrasse replied:
Done

DerekSnell commented on boards/arm/mimxrt1170_evk/Kconfig.defconfig:
```diff
@@ -68,9 +68,6 @@ if DISPLAY
 config MIPI_DSI
```
need to update copyright year 2023 in this file

DerekSnell commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk_cm7.dts:
```diff
@@ -39,13 +39,6 @@
 		device_type = "memory";
```
need to update copyright year 2023

DerekSnell commented on boards/arm/mimxrt595_evk/Kconfig.defconfig:
```diff
@@ -29,4 +29,16 @@ config HEAP_MEM_POOL_SIZE
 
```
need to update copyright year 2023

DerekSnell commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -33,6 +33,7 @@
 		zephyr,sram = &sram0;
```
need to update copyright year 2023

DerekSnell reviewed (changes requested):
Great work, @danieldegrasse .  Looks like a lot of work.

I do have some changes I would like to see before we merge this.  Thank you

danieldegrasse commented on samples/drivers/display/boards/mimxrt595_evk_cm33.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_HEAP_MEM_POOL_SIZE=1974272
+CONFIG_MCUX_LCDIF_EXTERNAL_FB_MEM=y
+# Use FlexSPI2 base address for framebuffer (pSRAM is present on this bus)
+CONFIG_MCUX_LCDIF_EXTERNAL_FB_ADDR=0x38000000
```
I think the way to do this might be to make these displays Zephyr shields, what are your thoughts on this approach? I can either integrate the change into this PR, or create a follow up PR to move the radium RM68200 display to be a shield, whichever you prefer

DerekSnell replied:
Hi @danieldegrasse ,
Creating shields makes sense to me.  A follow-up PR is fine with me, if that helps with getting this one approved.

danieldegrasse replied:
I'll create a follow up PR- this PR is already large, and I'd like to avoid adding additional changes to it that aren't required to get the hardware working

DerekSnell commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -170,6 +179,20 @@ static int memc_flexspi_init(const struct device *dev)
+		/* AHB master index, SOC specific */
+		flexspi_config.ahbConfig.buffer[i].masterIndex = data->buf_cfg[i].master_id;
+		/* RX buffer allocation (total available buffer space is instance/SOC specific) */
+		flexspi_config.ahbConfig.buffer[i].bufferSize = data->buf_cfg[i].buf_size;
```
nevermind, we will request this in the MCUXpresso SDK FlexSPI driver

danieldegrasse commented on dts/bindings/mipi-dsi/nxp,imx-mipi-dsi-host.yaml:
```diff
@@ -0,0 +1,79 @@
+#
+description: NXP MCUX MIPI DSI (Host architecture)
+
+compatible: "nxp,imx-mipi-dsi-host"
```
I've updated this binding name to `mipi-dsi-2l`, which is the best distinct name I can find for this IP block within our documentation

danieldegrasse commented on dts/bindings/display/nxp,lcdif.yaml:
```diff
@@ -0,0 +1,126 @@
+
+description: NXP LCDIF (LCD Interface) controller
+
+compatible: "nxp,lcdif"
```
I've renamed this binding to `nxp,dcnano-lcdif`, since this IP block is based on the DCNano LCDIF

danieldegrasse commented on samples/drivers/display/boards/mimxrt595_evk_cm33.conf:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_HEAP_MEM_POOL_SIZE=1974272
```
I've looked at this size a bit closer, and we only need a heap of roughly 33KB (32400 bytes for the display, and 128 bytes for kernel heap tracking data). I've updated the heap size accordingly.

DerekSnell commented on boards/arm/mimxrt595_evk/doc/index.rst:
```diff
@@ -103,6 +103,9 @@ already supported, which can also be re-used on this mimxrt595_evk board:
 | PM        | on-chip    | power management; uses SoC sleep,   |
 |           |            | deep sleep and deep-powerdown modes |
 +-----------+------------+-------------------------------------+
+| DISPLAY   | on-chip    | LCDIF; MIPI-DSI. Tested with RM68200|
```
Thank you for updating this doc page.  Can we also add to this doc NXP's module part number RK055HDMIPI4MA0 for this display?  And link to this product page https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/5-5-lcd-panel:RK055HDMIPI4MA0.  That way users can easily find this display.  The update does not need to be in this table, but somewhere in this doc.

danieldegrasse replied:
The display that we currently support (using the RM68200 controller) is the [RK055HDMIPI4M](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/5-5-lcd-panel:RK055HDMIPI4M). It uses a different controller than the updated RK055HDMIPI4MA0, as far as I can tell. I'll add a link to that display here

DerekSnell commented on drivers/display/Kconfig.mcux_dcnano_lcdif:
```diff
@@ -0,0 +1,56 @@
+	  be copied into the next one. This can have significant performance
+	  impact.
+
+config MCUX_DCNANO_LCDIF_MAINTAIN_CACHE
```
I'm adding my comment as another potential PR suggestion.  This MAINTAIN_CACHE concept is not specific to the LCDIF.  In this RT500 family, the cache is on the AHB slave port of the FlexSPI memory interface.  This means all masters accessing the FlexSPI share this common cache, and therefore coherency is not an issue.  Other drivers could use this feature as well.  Perhaps a better place for this feature is a DT property for the MEMC device.  And if the LCDIF or other drivers are accessing a MEMC device with MAINTAIN_CACHE property, then cache management is not required.  But we do not need to delay this PR to implement that.  Let's consider for a future improvement.

DerekSnell reviewed (dismissed):
@danieldegrasse ,
Thanks for all you work with this.  It looks great to me.  I just have a couple minor non-blocking comments

danieldegrasse commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -122,6 +123,7 @@
 		 * this state.
 		 */
 	};
+
```
Nope, removed it in latest push.

mmahadevan108 commented:
Please update the west file.

jfischer-no commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -0,0 +1,248 @@
+}
+
+#define MCUX_MIPI_DSI_DEVICE(id)								\
+	static const struct display_mcux_mipi_dsi_config display_mcux_mipi_dsi_config_##id = {	\
```
Please avoid this very long name, specially for variables with very limited scope, just
`static const struct mcux_mipi_dsi_config mipi_dsi_config_##id`.

danieldegrasse replied:
Fixed in lasted push

jfischer-no commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -0,0 +1,248 @@
+	.transfer = dsi_mcux_transfer,
+};
+
+static int display_mcux_mipi_dsi_init(const struct device *dev)
```
`static int mcux_mipi_dsi_init(const struct device *dev)`

jfischer-no commented on drivers/display/display_mcux_dcnano_lcdif.c:
```diff
@@ -0,0 +1,318 @@
+
+#ifdef CONFIG_MCUX_DCNANO_LCDIF_DOUBLE_FRAMEBUFFER
+	/* If display write is partial, copy current framebuffer to
+	 * queued one. Note- this has a significant
```
"Note- this has a significant" looks weird.

danieldegrasse replied:
Indeed it does:) I've fixed that comment to highlight what it was intending to, namely the potential performance implications of double framebuffering

jfischer-no commented on samples/drivers/display/boards/mimxrt595_evk_cm33.conf:
```diff
@@ -0,0 +1,16 @@
+#
```
We should always try to avoid board specific overlays because it adds maintenance overhead. We have cleaned up display controller samples a few times. Please revise if it can be done on the board level.

danieldegrasse replied:
I've moved the majority of this overlay into the board level- the only configuration I think should remain in this sample is the heap size, as this sample specifically allocates a large buffer from the heap for the display data. Other samples (like the LVGL demo) do not do this, and don't require additional heap size

jfischer-no reviewed (commented):
Looks good, just few comments.

danieldegrasse commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk_cm7.dts:
```diff
@@ -39,13 +39,6 @@
 		reg = <0x80000000 DT_SIZE_M(64)>;
 	};
-
-	en_mipi_display: enable-mipi-display {
```
https://github.com/zephyrproject-rtos/zephyr/pull/36875#discussion_r863932030 This seems to be the discussion regarding the regulator API usage, but I don't see a clear resolution in the comment chain. @mmahadevan108 Do you recall why a regulator node was used here?

gmarull replied:
Unless the display has a pin to control its power, we should stick with regulators. There's no guarantee that the display power rail won't be shared in other designs. If that's the case on any design, developers will need to fork and patch Zephyr, which isn't great.

danieldegrasse replied:
Understood, I've dropped the change that moved this power rail to be controlled by the display, so the regulator node is still present.

danieldegrasse committed:
drivers: kscan: fix GT911 support when not using interrupt

Fix support for GT911 touchscreen driver when not using interrupt.
The controller still requires that the INT pin be set to input in
order to correctly respond to I2C queries.

Also, the GT911 requires that the INT pin is not configured until the
RESET pin has been low at least 10 ms after reset, so add an additional
delay during initialization to account for this.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

jfischer-no commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -0,0 +1,248 @@
+};
+
+struct display_mcux_mipi_dsi_data {
+	const struct device *dev;
```
Looks like it is not used anywhere, also not initialized.

danieldegrasse replied:
You're right, not sure how this got missed. Thanks for catching it, I've remove this structure

jfischer-no commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -0,0 +1,248 @@
+	dsi_xfer.rxData = msg->rx_buf;
+
+	switch (msg->type) {
+
```
Stray empty line?

danieldegrasse replied:
Removed in latest push

jfischer-no commented on drivers/display/display_mcux_dcnano_lcdif.c:
```diff
@@ -0,0 +1,319 @@
+
+static void *mcux_dcnano_lcdif_get_framebuffer(const struct device *dev)
+{
+#ifdef CONFIG_MCUX_DCNANO_LCDIF_DOUBLE_FRAMEBUFFER
+	return NULL; /* Direct framebuffer access not supported */
+#else
+	const struct mcux_dcnano_lcdif_config *config = dev->config;
+
+	return config->fb_ptr;
+#endif
```
Seems like IS_ENABLED(CONFIG_MCUX_DCNANO_LCDIF_DOUBLE_FRAMEBUFFER) can be used here.
```c
	const struct mcux_dcnano_lcdif_config *config = dev->config;

	if (IS_ENABLED(CONFIG_MCUX_DCNANO_LCDIF_DOUBLE_FRAMEBUFFER)) {
		return config->fb_ptr;
	} else {
		return NULL;
	}

```

danieldegrasse replied:
Sure, I've changed this function to use `IS_ENABLED`

dleach02 commented on drivers/kscan/kscan_gt911.c:
```diff
@@ -242,7 +242,8 @@ static int gt911_init(const struct device *dev)
 	}
-
+	/* Delay at least 10 ms after power on before we configure gt911 */
+	k_sleep(K_MSEC(20));
```
why do we say "... delay at least 10 ms..." but then use 20msec instead of 10?

danieldegrasse replied:
I took this delay from our MCUXpresso SDK driver for the GT911, I assume that 20msec is used as a safe margin, since the GT911 datasheet requires at least 10msec of delay before any pins are configured after boot but does not impose a maximum delay
